### ISSUE TITLE
Long delay for webrequests

### ISSUE THREAD

**Issue by BaloMueller** [Azure/AKS#85]
Hello,

I'm using AKS to host a couple of Python/Flask applications. And I have serious performance issues.
The execution time (including db transactions, and everything else) is a according to newrelic at 200-500ms per request. Average 220ms.

But the time to first byte in my browser, when using https://www.webpagetest.org or `curl -o /dev/null -s -w %{time_total} http://` from several machines around the world, is always bigger 4s!

My service configuration:
```
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"backend","namespace":"default"},"spec":{"ports":[{"port":xxx}],"selector":{"app":"backend"},"type":"LoadBalancer"}}
  creationTimestamp: 2017-12-08T17:32:54Z
  name: backend
  namespace: default
  resourceVersion: "457699"
  selfLink: /api/v1/namespaces/default/services/backend
  uid: d2894312-dc3d-11e7-b51d-0a58ac1f0521
spec:
  clusterIP: 10.0.219.255
  externalTrafficPolicy: Cluster
  ports:
  - nodePort: 31395
    port: xxx
    protocol: TCP
    targetPort: xxx
  selector:
    app: backend
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - ip: 13.95.173.236
```

My deployment config:
```
apiVersion: extensions/v1beta1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "9"
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"apps/v1beta1","kind":"Deployment","metadata":{"annotations":{},"name":"backend","namespace":"default"},"spec":{"replicas":1,"template":{"metadata":{"labels":{"app":"backend"}},"spec":{"containers":[{"env":[xxx],"image":"xxx","livenessProbe":{"httpGet":{"httpHeaders":[{"name":"X-Custom-Header","value":"Awesome"}]},"initialDelaySeconds":3,"periodSeconds":30},"name":"backend","ports":[{"containerPort":xxx,"name":"web"}]}]}}}}
  creationTimestamp: 2017-12-08T17:32:54Z
  generation: 9
  labels:
    app: backend
  name: backend
  namespace: default
  resourceVersion: "862184"
  selfLink: /apis/extensions/v1beta1/namespaces/default/deployments/backend
  uid: d282326c-dc3d-11e7-b51d-0a58ac1f0521
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 2
  selector:
    matchLabels:
      app: backend-encryption
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app: backend
    spec:
      containers:
      - env:
           ... some env vars
        image: my image path
        imagePullPolicy: IfNotPresent
        name: backend-payment
        ports:
        - containerPort: xxxx
          name: web
          protocol: TCP
        resources: {}
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
      dnsPolicy: ClusterFirst
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: 2017-12-13T12:51:44Z
    lastUpdateTime: 2017-12-13T12:51:44Z
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  - lastTransitionTime: 2017-12-08T17:32:54Z
    lastUpdateTime: 2017-12-13T15:31:42Z
    message: ReplicaSet "backend-6bcdc6fb4f" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  observedGeneration: 9
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1
```

Here a waterfall screenshot from Chrome:
![screen shot 2017-12-19 at 10 21 25](https://user-images.githubusercontent.com/793558/34149807-9347b136-e4a6-11e7-8bfd-e27ecb20f8b9.png)

**Comment by slack** [comment]
Thanks for reaching out. The path for traffic to your application should be:
* Azure LB > AKS Nodes Availability Set > IP Tables Service Address > Pod

Few questions:
* How many nodes in your cluster?
* How many Pods for your app?
* Did you enable Azure LB Diagnostics to see LB/Node Health?

**Comment by brendandburns** [comment]
A few more things to try:

Launch a `busybox` container inside your cluster with:

`kubectl run busybox --rm -it --image=busybox`

Then run:

`time wget http://<service-ip-address>`

and

`time wget http://<pod-ip-address>`

That will help determine where the latency is coming from.

Thanks!

**Comment by BaloMueller** [comment]
Both of you: Thanks for your replies

@slack 
I agree that the problem is probably somewhere at the loadbalancer or the routing behind.
- Number of nodes: 1
- Pods per App: 1
- Just enabled LB diagnostics. As soon as I have some metrics, I'll post them here.

@brendandburns 
```
kubectl run busybox --rm -it --image=busybox
If you don't see a command prompt, try pressing enter.
time wget http://xxx/J14xxx
Connecting to encryption.rapitag.com (13.95.173.236:80)
real	0m 5.22s
user	0m 0.00s
sys	0m 0.00s
/ # time wget http://xxx/J14xxx
Connecting to encryption.rapitag.com (13.95.173.236:80)
wget: can't open 'J14xxx': File exists
Command exited with non-zero status 1
real	0m 3.17s
user	0m 0.00s
sys	0m 0.00s
/ # time wget http://xxx/J14xxx
Connecting to encryption.rapitag.com (13.95.173.236:80)
wget: can't open 'J14xxx': File exists
Command exited with non-zero status 1
real	0m 5.17s
user	0m 0.00s
sys	0m 0.00s
```

**Comment by BaloMueller** [comment]
@slack:
In 10 days, there were 2 Alerts about degraded LoadBalancer Health.
```
  | $table | AzureDiagnostics
-- | -- | --
  | TenantId | 76e8ee87-7edd-4768-bdc6-a692b9c97c7d
  | SourceSystem | Azure
  | TimeGenerated | 2017-12-31T14:12:01.552Z
  | Type | AzureDiagnostics
  | OperationName | LoadBalancerProbeHealthStatus
  | ResourceGroup | MC_RAPITAG-CONTAINER_RAPITAG-BACKEND_WESTEUROPE
  | SubscriptionId | 712aed80-8eec-4b16-8abf-4455246df58e
  | Category | LoadBalancerProbeHealthStatus
  | ResourceId | /SUBSCRIPTIONS/712AED80-8EEC-4B16-8ABF-4455246DF58E/RESOURCEGROUPS/MC_RAPITAG-CONTAINER_RAPITAG-BACKEND_WESTEUROPE/PROVIDERS/MICROSOFT.NETWORK/LOADBALANCERS/KUBERNETES
  | ResourceProvider | MICROSOFT.NETWORK
  | Resource | KUBERNETES
  | ResourceType | LOADBALANCERS
  | publicIpAddress_s | 52.174.189.1
  | systemId_g | de8fa58a-df4e-474a-9c32-499f74f13db9
  | port_d | 80
  | totalDipCount_d | 1
  | dipDownCount_d | 1
  | healthPercentage_d | 0
```

I only activated LB Diagnostics. Is there anything else I could activate to get better insights?

**Comment by slack** [comment]
@BaloMueller could you try connecting to the internal Kubernetes service name? Curious what those timings look like. All of the tests look like they are coming in through `13.95.173.236`, I'd expect to see a k8s service ip address `10.0.0.xxx` or similar.

**Comment by BaloMueller** [comment]
@slack Same problem there
```
Service/Loadbalancer internal IP:
/ # time wget http://10.0.219.255/J14xxx
Connecting to 10.0.219.255 (10.0.219.255:80)
J14xxx 100% |**************************************************************************************|    43   0:00:00 ETA
real	0m 5.14s
user	0m 0.00s
sys	0m 0.00s

Directly accessing the Pod Endoint:
/ # time wget http://10.244.0.37:3003/J14xxx
Connecting to 10.244.0.37:3003 (10.244.0.37:3003)
J14xxx 100% |**************************************************************************************|    43   0:00:00 ETA
real	0m 5.01s
user	0m 0.00s
sys	0m 0.00s

Though a traceroute is super fast (to both IPs):
traceroute to 10.244.0.37 (10.244.0.37), 30 hops max, 46 byte packets
 1  10.244.0.37 (10.244.0.37)  0.005 ms  0.003 ms  0.001 ms
```

I made this requests a couple of times. Both were always between 4.5s and 5.5s.

**Comment by brendandburns** [comment]
Two thoughts:
1) try direct access to the pod IP from the node running the pod that should eliminate azure network as a potential cause.

2) use tcpdump/Wireshark to analyze what is going on.

Is it possible your application is slow?

--brendan

________________________________

**Comment by slack** [comment]
Another thought, could be something in the app stack that is trying to do a reverse dns lookup. Consistently bad times ~ 5s makes me suspicious...

**Comment by BaloMueller** [comment]
I invested the whole day to investigate the problem. Now I'm even more confused.

I connected to the kubernetes agent and there into the docker container my app is running on using `docker exec -it 612701ab8a96 /bin/bash`. There I executed `'curl -o /dev/null -s -w "%{time_connect} + %{time_starttransfer} = %{time_total}\n" http://127.0.0.1:3001/xxx`. I did exactly the same on my local dev machine (same docker image, same MySQL Server which runs on Azure). Now the unexpected:

```
On AKS Agent:
Connection Time always less then 0.004s
Timetofirstbyte always bouncing between **3.5s and 4.5s**

Local Dev machine:
Connection Time always less then 0.006s
Timetofirstbyte consistent at **1.0s to 1.1s** (which is still too slow, I know)
```

So I thought: It must have something to do with the Mysql connection. I added timing logs into my application and measured the time each codeblock requires. I was surprised again:
```
before_request: 1515169051.46
end before_request: 0.00424218177795
mycode_invoked: 0.00434017181396
Got_db_cursor: 0.0102941989899
db_query_executed: 0.0137300491333
db_result_fetched: 0.0138421058655
request_teardown_end: 0.0160691738129
```
And these numbers are pretty much identical on my local machine and on the kubernetes node.
I saw requests that finished the [teardown](http://flask.pocoo.org/docs/0.12/api/#flask.Flask.do_teardown_request) in 0.015s, but the TTFB using curl within the same container were at 3.8s.

Conclusion: Same code, same container, same DB Server, same app execution times, but different TTFB from within the same container. Any suggestions how that's even possible?

**Comment by BaloMueller** [comment]
I terminated the Kubernetes Cluster and created a new one. There the problem doesn't exist. Everything is smooth and fast.

**Comment by invidious9000** [comment]
This is still an issue, out-of-the-box AKS clusters in multiple regions have sporadic traffic hiccups where it consistently adds exactly 5.5s to transit time.

**Comment by posix4e** [comment]
Same this is destroying us

**Comment by invidious9000** [comment]
I've opened another ticket under https://github.com/Azure/AKS/issues/632 that I think may be the same underlying issue as this one.

**Comment by posix4e** [comment]
We fixed this by switch dns to tcp

**Comment by TiagoJMartins** [comment]
@posix4e Could you please elaborate on your solution?

**Comment by yangl900** [comment]
This issue is pretty old, but want to give an update since I happened to look at another very similar customer report.

I believe the original issue reported by @BaloMueller was due to single threaded Flask + Azure LB probe. The up to 5 seconds latency is caused by the TCP probe from load balancer. When Azure LB does a probe, it will initiate a tcp connection to the application, however it will not send any http headers. Flask, on the other hand, is single threaded and will wait for the http headers and while it is waiting, other request will be queued. This will cause the unexpected latency.

To mitigate this, either enable multi-threading for the application, or, set the `externalTrafficPolicy: Local` on the svc. With the external traffic local, the LB health probe will be handled by a node port instead of the application port. The issue should be gone after that.
